infrahub_sdk/spec/object.py

```diff
@@ -0,0 +1,128 @@
+                    )
+                    # clean_data[key] = value[data]
+                remaining_rels.append(key)
+
+        if context:
+            clean_context = {
+                ckey: cvalue
+                for ckey, cvalue in context.items()
+                if ckey in schema.relationship_names + schema.attribute_names
+            }
+            clean_data.update(clean_context)
+
+        clean_data = cls.enrich_node(data=clean_data, context=context or {})
+
+        node = await client.create(kind=schema.kind, branch=branch, data=clean_data)
+        await node.save(allow_upsert=True)
+        display_label = node.get_human_friendly_id_as_string() or f"{node.get_kind()} : {node.id}"
+        client.log.info(f"Node: {display_label}")
+
+        for rel in remaining_rels:
+            # identify what is the name of the relationship on the other side
+            if not isinstance(data[rel], dict) and "data" in data[rel]:
+                raise ValueError(f"relationship {rel} must be a dict with 'data'")
+
+            rel_schema = schema.get_relationship(name=rel)
+            peer_kind = data[rel].get("kind", default_schema_kind) or rel_schema.peer
+            peer_schema = await client.schema.get(kind=peer_kind, branch=branch)
+
+            if rel_schema.identifier is None:
+                raise ValueError("identifier must be defined")
+
+            peer_rel = peer_schema.get_relationship_by_identifier(id=rel_schema.identifier)
+
+            rel_data = data[rel]["data"]
+            context = {}
+            if peer_rel:
+                context[peer_rel.name] = node.id
+
+            if rel_schema.cardinality == "one" and isinstance(rel_data, dict):
+                await cls.create_node(
+                    client=client,
+                    schema=peer_schema,
+                    data=rel_data,
+                    context=context,
+                    branch=branch,
+                    default_schema_kind=default_schema_kind,
+                )
+
+            elif rel_schema.cardinality == "many" and isinstance(rel_data, list):
+                for idx, peer_data in enumerate(rel_data):
+                    context["list_index"] = idx
+                    await cls.create_node(
+                        client=client,
+                        schema=peer_schema,
+                        data=peer_data,
+                        context=context,
+                        branch=branch,
+                        default_schema_kind=default_schema_kind,
+                    )
+            else:
+                raise ValueError(
+                    f"Relationship {rel_schema.name} doesn't have the right format {rel_schema.cardinality} / {type(rel_data)}"
+                )
+
+
+class ObjectFile(InfrahubFile):
+    _spec: Optional[InfrahubObjectFileData] = None
+
+    @property
+    def spec(self) -> InfrahubObjectFileData:
```

Comment:
Just thinking out loud here if we're later adding a spec property to other object types and want to look at the spec for a generic `InfrahubFile` object it might get problematic if we always need to override the returned data in a different way for each given subclass (in this case `ObjectFile` which would return an `InfrahubObjectFileData`. Depending on what we do in the future it might not be a problem, alternatively we could look at generics and see if we could fix it with that.

Reply:
Agreed, not a fan of this implementation as well, happy to consider a better/cleaner solution in the future